src/database-postgresql.cpp

```diff
@@ -0,0 +1,238 @@
+{
+	prepareStatement("read_block",
+			"SELECT block_data FROM blocks "
+					"WHERE posX = $1::int AND posY = $2::int AND posZ = $3::int");
+
+	prepareStatement("write_block",
+			"INSERT INTO blocks (posX, posY, posZ, block_data) VALUES "
+					"($1::int, $2::int, $3::int, $4::bytea) "
+					"ON CONFLICT ON CONSTRAINT blocks_pkey DO UPDATE SET block_data = $4::bytea");
+
+	prepareStatement("delete_block", "DELETE FROM blocks WHERE posX = $1::int AND posY = $2::int AND posZ = $3::int");
+
+	prepareStatement("list_all_loadable_blocks", "SELECT posX, posY, posZ FROM blocks");
+}
+
+void Database_PostgreSQL::prepareStatement(const std::string &name, const std::string &sql)
+{
+	resultsCheck(PQprepare(m_conn, name.c_str(), sql.c_str(), 0, NULL));
+}
+
+PGresult* Database_PostgreSQL::resultsCheck(PGresult* result, bool clear /*= true*/)
+{
+	ExecStatusType statusType = PQresultStatus(result);
+
+	if ((statusType != PGRES_COMMAND_OK && statusType != PGRES_TUPLES_OK) ||
+		statusType == PGRES_FATAL_ERROR) {
+		throw DatabaseException(std::string(
+			"PostgreSQL database error: ") +
+			PQresultErrorMessage(result));
+	}
+
+	if (clear)
+		PQclear(result);
+
+	return result;
+}
+
+void Database_PostgreSQL::createDatabase()
+{
+	PGresult *result = resultsCheck(PQexec(m_conn,
+		"SELECT relname FROM pg_class WHERE relname='blocks';"),
+		false);
+
+	// If table doesn't exist, create it
+	if (!PQntuples(result)) {
+		static const char* dbcreate_sql = "CREATE TABLE blocks ("
+			"posX INT NOT NULL,"
+			"posY INT NOT NULL,"
+			"posZ INT NOT NULL,"
+			"block_data BYTEA,"
+			"PRIMARY KEY (posX,posY,posZ)"
+		");";
+		resultsCheck(PQexec(m_conn, dbcreate_sql));
+	}
+
+	PQclear(result);
+
+	infostream << "PostgreSQL: Game Database was inited." << std::endl;
+}
+
+
+void Database_PostgreSQL::beginSave()
+{
+	verifyDatabase();
+	resultsCheck(PQexec(m_conn, "BEGIN;"));
+}
+
+void Database_PostgreSQL::endSave()
+{
+	resultsCheck(PQexec(m_conn, "COMMIT;"));
+}
+
+bool Database_PostgreSQL::saveBlock(const v3s16 &pos, const std::string &data)
+{
+	verifyDatabase();
+
+	// This case is manually handled because we manipulate blob datas
+	std::string x = itos(pos.X), y = itos(pos.Y), z = itos(pos.Z);
+	const char *values[] = {x.c_str(), y.c_str(), z.c_str(), data.c_str()};
+	const int valLen[] = {sizeof(int), sizeof(int), sizeof(int), (int)data.size()};
+	const int valFormats[] = { 0, 0, 0, 1 };
+
+	resultsCheck(PQexecPrepared(m_conn, "write_block", 4,
+				values, valLen, valFormats, 1), true);
+	return true;
+}
+
+void Database_PostgreSQL::loadBlock(const v3s16 &pos, std::string *block)
+{
+	verifyDatabase();
+
+	std::string x = itos(pos.X), y = itos(pos.Y), z = itos(pos.Z);
+	const char *values[] = {x.c_str(), y.c_str(), z.c_str()};
+
+	PGresult *results = execPrepared("read_block", 3, values, false, false);
+
+	*block = "";
+
+	if (PQntuples(results)) {
+		size_t rd_s;
+		unsigned char* rd = PQunescapeBytea((unsigned char*)PQgetvalue(results, 0, 0), &rd_s);
```

Comment:
Right but if you look at the documentation for PQunescapeBytesa you'll see that it's only needed when handling data in text mode instead of binary mode.  If you were using binary mode to begin with, the official documentation states that you wouldn't need this step.

Please review the documentation for PostgreSQL's C/++ interface.